jetty-util/src/main/java/org/eclipse/jetty/util/security/CredentialFactory.java

```diff
@@ -0,0 +1,68 @@
+//
+//  ========================================================================
+//  Copyright (c) 1995-2016 Mort Bay Consulting Pty. Ltd.
+//  ------------------------------------------------------------------------
+//  All rights reserved. This program and the accompanying materials
+//  are made available under the terms of the Eclipse Public License v1.0
+//  and Apache License v2.0 which accompanies this distribution.
+//
+//      The Eclipse Public License is available at
+//      http://www.eclipse.org/legal/epl-v10.html
+//
+//      The Apache License v2.0 is available at
+//      http://www.opensource.org/licenses/apache2.0.php
+//
+//  You may elect to redistribute this code under either of these licenses.
+//  ========================================================================
+//
+
+package org.eclipse.jetty.util.security;
+
+import java.util.ServiceLoader;
+
+import org.eclipse.jetty.util.security.Credential.Crypt;
+import org.eclipse.jetty.util.security.Credential.MD5;
+
+/**
+ * Class responsible for the creation of authentication credentials from strings.
+ * It supports both the known Credential types (CRYPT and MD5) and some 
+ * Credential types discovered at runtime by using the service loader mechanism.
+ *
+ * @see Credential
+ * @see Password
+ */
+public class CredentialFactory
+{
+	private ServiceLoader<CredentialProvider> credentialProviderLoader;
+
+	/**
+	 * Constructor
+	 */
+    public CredentialFactory() {
+        super();
+        credentialProviderLoader = ServiceLoader.load(CredentialProvider.class);
+    }
+
+    /**
+     * Get a credential from a String. If the credential String starts with a
+     * known Credential type (eg "CRYPT:" or "MD5:" ) then a Credential of that
+     * type is returned. Otherwise, it tries to find a credential provider 
+     * whose prefix matches with the start of the credential String. Else the 
+     * credential is assumed to be a Password.
+     * 
+     * @param credential String representation of the credential
+     * @return A Credential or Password instance.
+     */
+    public Credential getCredential(String credential) {
+    	if (credential.startsWith(Crypt.__TYPE)) return new Crypt(credential);
+        if (credential.startsWith(MD5.__TYPE)) return new MD5(credential);
+    	
+        for (CredentialProvider cp : credentialProviderLoader) {
+            if (credential.startsWith(cp.getPrefix())) {
+                return cp.getCredential(credential);
```

Comment:
perhaps allow for a null return here and then to keep trying other providers, so you can have multiple providers for the same prefix.